lib/Sema/TypeCheckStmt.cpp

```diff
@@ -442,6 +442,32 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
       return RS;
     }
 
+    // If the body consisted of a single return without a result
+    // 
+    //   func foo() -> Int {
+    //     return
+    //   }
+    // 
+    // in parseAbstractFunctionBody the return is given an empty, implicit tuple
+    // as its result
+    //
+    //   func foo() -> Int {
+    //     return ()
+    //   }
+    //
+    // Look for that case and diagnose it as missing return expression.
+    if (!ResultTy->isVoid() && TheFunc->hasSingleExpressionBody()) {
+      auto expr = TheFunc->getSingleExpressionBody();
+      if (expr->isImplicit()) {
+        if (auto tuple = dyn_cast<TupleExpr>(expr)) {
+          if (tuple->getNumElements() == 0 && !ResultTy->isVoid()) {
```

Comment:
You already checked `ResultTy->isVoid()` at the top level. (It might also make sense to merge this with the previous `if`, since they're checking related conditions.)

Reply:
Fixed in https://github.com/apple/swift/pull/25802/commits/3761b02d74721cf0ccedae16ba17c2500ef5637a .  Is 

```
if (expr->isImplicit() &&
  dyn_cast<TupleExpr>(expr) != nullptr &&
  dyn_cast<TupleExpr>(expr)->getNumElements() == 0) {
```

what you had in mind for merging the ifs together?

Reply:
I actually meant

```
if (ResultTy->isVoid()) {
  if (!RS->hasResult()) { /* old stuff */ }
  if (TheFunc->hasSingleExpressionBody()) { /* new stuff */ }
}
```

Up to you, though.

Reply:
If I'm understanding this right, that wouldn't quite work because the old stuff

```
if (!RS->hasResult()) {
  if (!ResultTy->isVoid())
    TC.diagnose(RS->getReturnLoc(), diag::return_expr_missing);
  return RS;
}
```

returns `RS` always whenever `!RS->hasResult()` so there'd have to be an additional check after the `if (ResultTy->isVoid()) {...}` for `if (!RS->hasResult())`.

Reply:
Oops, you're right. I didn't notice that the return wasn't nested.